File: LongestRepeatingCharReplacement.java
// https://leetcode.com/problems/longest-repeating-character-replacement/
/*
Approach -
end-start+1 = size of the current window, maxCount = largest count of a unique char in the current window
When end-start+1-maxCount == 0, then then the window is filled with only one character
When end-start+1-maxCount > 0, then we have characters in the window that are not the char that occurs the most.
end-start+1-maxCount is equal to exactly the # of characters that are NOTnot the char that occurs the most in that window.
We are allowed to have at most k replacements in the window, so when end-start+1-maxCount > k, 
then there are more characters in the window than we can replace, and we need to shrink the window.
*/
class Solution {
     public int characterReplacement(String s, int k) {
        int len = s.length();
        int[] count = new int[26];
        int start = 0, maxCount = 0, maxLength = 0;
        for (int end = 0; end < len; end++) {
            maxCount = Math.max(maxCount, ++count[s.charAt(end) - 'A']);
            if (end - start + 1 - maxCount > k) {
                count[s.charAt(start) - 'A']--;
                start++;
            }
            maxLength = Math.max(maxLength, end - start + 1);
        }
        return maxLength;
    }
}

File: LongestRepeatingCharReplacement.py
# https://leetcode.com/problems/longest-repeating-character-replacement/
# Binary search + sliding window
class Solution(object):
    def characterReplacement(self, s, k):
        """
        :type s: str
        :type k: int
        :rtype: int
        """
        lo = 1
        hi = len(s)+1
        while lo!=hi-1:
            mid = (lo+hi)/2
            if self.makeValidSubstring(s, mid, k):
                lo = mid
            else:
                hi = mid
        return lo

    def makeValidSubstring(self, s, mid, k):
        freq_map = {}
        max_frequency = 0
        start = 0
        for end in range(len(s)):
            freq_map[s[end]] = freq_map.get(s[end], 0) + 1
            if end + 1 - start > mid:
                freq_map[s[start]] -= 1
                start += 1
            max_frequency = max(max_frequency, freq_map[s[end]])
            if mid - max_frequency <= k:
                return True
        return False